### PR TITLE
[WIP] thread safety: use osgDB::readRef*File()

### PR DESCRIPTION
**Pull request by filnet**
New read*File() methods are in osgEarth/Utils

The read*File() methods could stay in Utils or be moved to a new osgEarth/FileRead.

### CODE REVIEW

File: src/osgEarth/OverlayDecorator.cpp [hunk 1]
```diff
@@ -722,17 +723,17 @@ OverlayDecorator::cullTerrainAndCalculateRTTParams(osgUtil::CullVisitor* cv,
             {
                 frustumPH.dumpGeometry(0,0,0,fn);
             }
-            osg::Node* camNode = osgDB::readNodeFile(fn);
+            osg::ref_ptr<osg::Node> camNode = readNodeFile(fn);
```
Comment (filnet) [inline src/osgEarth/OverlayDecorator.cpp]: change of variable type needs to be reviewed.

File: src/osgEarth/URI.cpp [hunk 2]
```diff
@@ -300,7 +300,7 @@ namespace
             }
             return HTTPClient::readObject(req, opt, p);
         }
-        ReadResult fromFile( const std::string& uri, const osgDB::Options* opt ) { return ReadResult(osgDB::readObjectFile(uri, opt)); }
+        ReadResult fromFile( const std::string& uri, const osgDB::Options* opt ) { return ReadResult(readObjectFile(uri, opt)); }
```
Comment (filnet) [inline src/osgEarth/URI.cpp]: ReadResult has a constructor that takes osg::ref_ptr<>

File: src/osgEarth/Utils [hunk 3]
```diff
@@ -22,11 +22,18 @@
 #ifndef OSGEARTH_UTILS_H
 #define OSGEARTH_UTILS_H 1
 
+#include <osg/Version>
+
+#if OSG_VERSION_GREATER_OR_EQUAL(3,5,1)
+#undef OSG_PROVIDE_READFILE
```
Comment (filnet) [inline src/osgEarth/Utils]: note this trick to make sure osgDB::read*File() methods are not visible.

File: src/osgEarth/Utils [hunk 4]
```diff
@@ -252,6 +259,129 @@ namespace osgEarth
         virtual ~AllocateAndMergeBufferObjectsVisitor() { }
         void apply(osg::Geode& geode);
     };
+
+    inline osg::ref_ptr<osg::Object> readObjectFile(const std::string& filename, const osgDB::Options* options)
+    {
+#if OSG_VERSION_GREATER_OR_EQUAL(3,4,0)
```
Comment (filnet) [inline src/osgEarth/Utils]: could be 2.3.0...

File: src/osgEarth/Utils [hunk 5]
```diff
@@ -22,11 +22,18 @@
...
+#if OSG_VERSION_GREATER_OR_EQUAL(3,5,1)
+#undef OSG_PROVIDE_READFILE
+#endif
+
 #include <osgEarth/Common>
 #include <osgEarth/StringUtils>
 
 #include <osg/Vec3f>
 #include <osg/AutoTransform>
+#include <osgDB/ReadFile>
```
Comment (filnet) [inline src/osgEarth/Utils]: this is the *one and only* place where osgDB/ReadFile is included.

File: src/osgEarth/Utils [hunk 6]
```diff
@@ -252,6 +259,129 @@ namespace osgEarth
...
+#else
+        return osgDB::readObjectFile(filename);
+#endif
+    }
+
+    template<typename T>
+    inline osg::ref_ptr<T> readFile(const std::string& filename, const osgDB::Options* options)
+    {
+#if OSG_VERSION_GREATER_OR_EQUAL(3,5,1)
+        return osgDB::readRefFile<T>(filename, options);
```
Comment (filnet) [inline src/osgEarth/Utils]: readRefFile (a templatized version of readRefObjectFile) was introduced in osg 3.5.1.

File: src/osgEarthDrivers/bing/BingTileSource.cpp [hunk 7]
```diff
@@ -123,7 +124,7 @@ class BingTileSource : public TileSource, public osgEarth::Drivers::BingOptions
      */
     osg::Image* createImage( const TileKey& key, ProgressCallback* progress )
     {
-        osg::Image* image = 0L;
+        osg::ref_ptr<osg::Image> image;
```
Comment (filnet) [inline src/osgEarthDrivers/bing/BingTileSource.cpp]: change of variable type needs to be reviewed

File: src/osgEarthDrivers/sky_simple/SimpleSkyNode.cpp [hunk 8]
```diff
@@ -615,7 +615,7 @@ SimpleSkyNode::makeMoon()
     osg::Geometry* geom = s_makeEllipsoidGeometry( em.get(), em->getRadiusEquator(), true );    
     //TODO:  Embed this texture in code or provide a way to have a default resource directory for osgEarth.
     //       Right now just need to have this file somewhere in your OSG_FILE_PATH
-    osg::Image* image = osgDB::readImageFile( "moon_1024x512.jpg" );
+    osg::ref_ptr<osg::Image> image = readImageFile( "moon_1024x512.jpg" );
```
Comment (filnet) [inline src/osgEarthDrivers/sky_simple/SimpleSkyNode.cpp]: change of variable type needs to be reviewed.

File: src/osgEarthUtil/TileIndexBuilder.cpp [hunk 9]
```diff
@@ -29,7 +29,6 @@
 #include <osgDB/FileUtils>
 #include <osgDB/FileNameUtils>
 
-using namespace osgDB;
```
Comment (filnet) [inline src/osgEarthUtil/TileIndexBuilder.cpp]: that was the only place where osgDB namespace was used...

**Comment by filnet** [comment]
The PR had some interesting cleanups that are not in master.
Will try to rework this PR to make it smaller and more manageable.

**Comment by gwaldron** [comment]
Sorry -- I thought we had incorporated all this! Will reopen.

**Comment by filnet** [comment]
This PR goes a bit further (too much I think...) than what we did together.

It adds an osgEarth wrapper class around osgDB::readFile() methods.
Let me know if you find the approach interesting or overkill.
If you like it, then I'll fix the conflicts and polish the PR.

File: src/osgEarth/Units [hunk 10]
```diff
@@ -23,6 +23,14 @@
 #include <osgEarth/Config>
 #include <ostream>
 
+#define GCC_VERSION (__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__)
+
+#if GCC_VERSION >= 50300
+#define OPTIMIZE __attribute__((optimize("no-ipa-sra")))
+#else
+#define OPTIMIZE
+#endif
```
Comment (filnet) [inline src/osgEarth/Units]: spurious commit, need to go.

**Comment by gwaldron** [comment]
OBE - resolved in other ways

**Pull request closed without merging**